### DISCUSSION TITLE
GraphQL to support federation

### DISCUSSION THREAD

**Discussion by jaspertandy** [craftcms/cms#8853] [Ideas]
I was torn on whether to post this as a bug, but since the GraphQL implementation doesn't claim to support federation, it's not.

However, it would make Craft's GraphQL implementation so useful to support the [federation spec](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/). Is this on the roadmap at all? I have a project coming up that would hugely benefit from it if it's at all possible.

At the moment, when I try to set Craft as a downstream in a federated config, I receive the following error:

```
Error checking for changes to service definitions: Couldn't load service definitions for "craft" at http://craft/graphql: Cannot query field "_service" on type "Query".
```

```
const gateway = new ApolloGateway({
  serviceList: [
    { name: 'users', url: 'http://users:4000' },
    { name: 'collections', url: 'http://collections:4000' },
    { name: 'craft', url: 'http://craft/graphql' }
  ]
});
```

This pointed me to [this issue](https://github.com/apollographql/apollo-server/issues/3419) on Apollo Server, which then links off to the federation spec.

Would this be something that's feasible to add?

**Comment by andris-sevcenko** [comment]
It would need to be supported by the library that we're using, but I'm not sure if that's on a workable timetable, there. There was some work going on regarding that at some point, but it seems like it stalled. (https://github.com/nuwave/lighthouse/pull/1051)

I'll keep this open, though, as it is a useful feature and maybe we can figure out a way of adding it.

**Comment by charlie-hadden** [comment]
Just to note: it appears that there is at least experimental support for this upstream now: https://lighthouse-php.com/5/federation/getting-started.html

**Comment by marjorg** [comment]
@andris-sevcenko Would be cool to have a way to plug into this using Craft

**Comment by charlie-hadden** [comment]
After posting that link yesterday I started to have a bit of a deeper look at this. It seems that lighthouse isn't currently used so I presume craft has migrated away from it since it was first brought up here.

As an alternative I started to look into what it would take to add support with a plugin. I've made a little bit of progress on that route already. As long as I don't run into any major issues I think I should be able to get a rough prototype working fairly soon. The code quality isn't great at the moment - partly because it's just a quick proof of concept, and partly because of some limitations with the events that craft triggers. Hopefully it can become a decent starting point though. If I'm able to get it to a suitable point then I'll get the code posted up here.

**Reply by andris-sevcenko** [reply]
Lighthouse was never used by Craft - they just also use the same library we're using under the hood, so, if Lighhouse supports it, there might be a way for us to support it, too.

**Reply by charlie-hadden** [reply]
Ah - that makes sense. I'd misunderstood your original message as I hadn't looked into it very deeply when this issue was first posted up here.

**Comment by charlie-hadden** [comment]
I've now got a proof of concept available here if anybody wants to check it out: https://github.com/builtbybuffalo/craft-apollo-federation

At the moment it's very much a proof of concept. It could probably be made quite a bit more efficient (the `_entities` resolver currently just resolves a single entry at a time), and I haven't tested how it works with multiple schemas or craft's GraphQL access tokens yet. It also doesn't allow for craft to reference other types in the full graph.

What it _does_ allow, though, is referring to craft types from other sub-graphs. So a schema such as this would be valid for another service:

```gql
extend type blogPosts_blogPosts_Entry {
  id: ID @external
}

type Product {
  id: ID!
  featuredBlogPosts: [blogPosts_blogPosts_Entry]
}

type Query {
  products: [Product]
}
```

Whilst setting up the plugin I took a look over the code and cleaned it up a little bit, so I'm actually not too concerned about the events that craft exposes any more.

We'll be making use of this on a project we're starting so I'll try and make sure any improvements we make to it end up back in that repository. I'll be focusing mainly on the features that we need, but if it ends up being useful then we can look at publishing it properly.

**Comment by andris-sevcenko** [comment]
Just for an update to this: since we use [webonyx/graphql-php](https://github.com/webonyx/graphql-php/) package, which follows the [reference implementation ](https://github.com/graphql/graphql-js), it is extremely unlikely for us to divert and start implementing features, that are not part of the specification.

Doing so would be rather irresponsible, as we'd be in a pickle once a said feature appears in the specification since it's very likely the implementation would be different from ours and we'd be stuck maintaining the same functionality with two different pieces of code or forcing a painful migration on everybody relying on the bootleg implementation.

**Reply by charlie-hadden** [reply]
I'm not sure how likely it would be that federation would be made part of the main GraphQL specification so I'm not sure how much of a concern that would be in practice. The federation spec is basically just a thin layer of extra functionality on top of the main GraphQL specification.

Being that it's separate from the main specification though, I can certainly see the flip side and a good argument that it perhaps shouldn't be a part of the craft core. The plugin I started work on and mentioned in a different comment here seems like a good compromise - it should cover our particular use case well, and be easy to swap out later on if situations change. Since working on that I've come to the conclusion that it's likely best suited as a plugin anyway.

**Reply by andris-sevcenko** [reply]
Beautiful, can't wait to see the final result, once it's all polished and good to go!
